Sources/_MatchingEngine/Regex/Parse/LexicalAnalysis.swift

```diff
@@ -456,11 +456,108 @@ extension Source {
     return AST.Trivia(trivia)
   }
 
+  /// Try to lex a matching option.
+  ///
+  ///     MatchingOption -> 'i' | 'J' | 'm' | 'n' | 's' | 'U' | 'x' | 'xx' | 'w'
+  ///                     | 'D' | 'P' | 'S' | 'W' | 'y{' ('g' | 'w') '}'
+  ///
+  mutating func lexMatchingOption() throws -> AST.MatchingOption? {
+    typealias OptKind = AST.MatchingOption.Kind
+
+    let locOpt = try recordLoc { src -> OptKind? in
+      func advanceAndReturn(_ o: OptKind) -> OptKind {
+        src.advance()
+        return o
+      }
+      guard let c = src.peek() else { return nil }
+      switch c {
+      // PCRE options.
+      case "i": return advanceAndReturn(.caseInsensitive)
+      case "J": return advanceAndReturn(.allowDuplicateGroupNames)
+      case "m": return advanceAndReturn(.multiline)
+      case "n": return advanceAndReturn(.noAutoCapture)
+      case "s": return advanceAndReturn(.singleLine)
+      case "U": return advanceAndReturn(.reluctantByDefault)
+      case "x":
+        src.advance()
+        return src.tryEat("x") ? .extraExtended : .extended
+
+      // ICU options.
+      case "w": return advanceAndReturn(.unicodeWordBoundaries)
+
+      // Oniguruma options.
+      case "D": return advanceAndReturn(.asciiOnlyDigit)
+      case "P": return advanceAndReturn(.asciiOnlyPOSIXProps)
+      case "S": return advanceAndReturn(.asciiOnlySpace)
+      case "W": return advanceAndReturn(.asciiOnlyWord)
+      case "y":
+        src.advance()
+        try src.expect("{")
+        let opt: OptKind
+        if src.tryEat("w") {
+          opt = .textSegmentWordMode
+        } else {
+          try src.expect("g")
+          opt = .textSegmentGraphemeMode
+        }
+        try src.expect("}")
+        return opt
+
+      default:
+        return nil
```

Comment:
We have a known terminator (`:` or `)`), so it seems like we'd want to gather options until we hit the terminator (and error if we encounter anything not an option) and error if we reach end-of-input as well. I think that could also simplify some of this logic, as it's fine to consume a non-terminating character up front.

Reply:
Hmm it's tricky because there's also other valid group syntax that starts with `(?` (even containing letter characters for e.g `(?P<`, unless we guaranteed we would have already tried lexing them before trying to lex matching options? Or otherwise make sure to exclude the other valid group syntax here?